File: internal/ssh_config.go
package ll

import (
	"os"
	"path/filepath"

	"github.com/kevinburke/ssh_config"
	"github.com/yuin/gopher-lua"
)

func sshconfigPort(L *lua.LState) int {
	host := L.CheckString(1)
	port := ssh_config.Get(host, "Port")
	if port != "" {
		L.Push(lua.LString(port))
		return 1
	} else {
		L.Push(lua.LNil)
		L.Push(lua.LString("ssh_config: No such host."))
		return 2
	}

}

func sshconfigIdentityFile(L *lua.LState) int {
	host := L.CheckString(1)
	key := ssh_config.Get(host, "IdentityFile")
	if key != "" {
		L.Push(lua.LString(key))
		return 1
	} else {
		L.Push(lua.LNil)
		L.Push(lua.LString("ssh_config: No such host."))
		return 2
	}
}

func sshconfigHostname(L *lua.LState) int {
	host := L.CheckString(1)
	hn := ssh_config.Get(host, "Hostname")
	if hn != "" {
		L.Push(lua.LString(hn))
		return 1
	} else {
		L.Push(lua.LNil)
		L.Push(lua.LString("ssh_config: No such host."))
		return 2
	}
}

func sshconfigHosts(L *lua.LState) int {
	f, _ := os.Open(filepath.Join(os.Getenv("HOME"), ".ssh", "config"))
	c, _ := ssh_config.Decode(f)
	t := L.NewTable()
	for _, host := range c.Hosts {
		for _, pat := range host.Patterns {
			s := pat.String()
			if s != "*" {
				t.Append(lua.LString(s))
			}
		}
	}
	L.Push(t)
	return 1
}

func SSHconfigLoader(L *lua.LState) int {
	t := L.NewTable()
	L.SetFuncs(t, sshconfigApi)
	L.Push(t)
	return 1
}

var sshconfigApi = map[string]lua.LGFunction{
	"port":          sshconfigPort,
	"identity_file": sshconfigIdentityFile,
	"hostname":      sshconfigHostname,
	"hosts":         sshconfigHosts,
}

File: tests/ssh_config.lua
local included = pcall(debug.getlocal, 4, 1)
local T = require("test")
local ssh_config = require("ssh_config")
--# = ssh_config
--# :toc:
--# :toc-placement!:
--#
--# Get values from ~/.ssh/config
--#
--# toc::[]
--#
--# == *ssh_config.port*(_String_) -> _String_
--# Get configured Port for Host.
--#
--# === Arguments
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |string |Host
--# |===
--#
--# === Returns
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |string |Port
--# |===
local ssh_config_port = function()
	T.is_function(ssh_config.port)
end
--#
--# == *ssh_config.hostname*(_String_) -> _String_
--# Get configured Hostname for Host.
--#
--# === Arguments
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |string |Host
--# |===
--#
--# === Returns
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |string |Hostname
--# |===
local ssh_config_hostname = function()
	T.is_function(ssh_config.hostname)
end
--#
--# == *ssh_config.identity_file*(_String_) -> _String_
--# Get configured IdentityFile for Host.
--#
--# === Arguments
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |string |Host
--# |===
--#
--# === Returns
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |string |Path of key
--# |===
local ssh_config_identityfile = function()
	T.is_function(ssh_config.identity_file)
end
--#
--# == *ssh_config.hosts*() -> _Table_
--# Get all hosts configured in ~/.ssh/config
--#
--# === Returns
--# [options="header",width="72%"]
--# |===
--# |Type |Description
--# |table|List of hosts
--# |===
local ssh_config_hosts = function()
	T.is_function(ssh_config.hosts)
	T.is_table(ssh_config.hosts())
end
if included then
	return function()
		T["ssh_config.port"] = ssh_config_port
		T["ssh_config.hostname"] = ssh_config_hostname
		T["ssh_config.identityfile"] = ssh_config_identityfile
		T["ssh_config.hosts"] = ssh_config_hosts
	end
else
	T["ssh_config.port"] = ssh_config_port
	T["ssh_config.hostname"] = ssh_config_hostname
	T["ssh_config.identityfile"] = ssh_config_identityfile
	T["ssh_config.hosts"] = ssh_config_hosts
end
